extensions/amp-analytics/0.1/transport.js

```diff
@@ -118,13 +124,194 @@ export class Transport {
+   * of the amp-analytics config object
+   * @param {Function} processResponse An optional function to receive any
+   * response messages back from the cross-domain iframe
+   */
+  static processCrossDomainIframe(ampDoc, transportOptions, processResponse) {
+    user().assert(!(transportOptions['beacon'] || transportOptions['xhrpost'] ||
+      transportOptions['image']), 'Cross-domain frame cannot coexist with' +
+      ' other transport methods');
+    const frameUrl = Transport.appendHashToUrl_(transportOptions['iframe'],
+        transportOptions['dataHash']);
+    // If iframe doesn't exist for this iframe url (and data hash), create it.
+    if (Transport.hasCrossDomainFrame(frameUrl)) {
+      this.incrementUsageCount_(frameUrl);
+    } else {
+      const frame = Transport.createCrossDomainFrame(ampDoc, frameUrl);
+      ampDoc.body.appendChild(frame);
+    }
+    if (processResponse) {
+      window.addEventListener("message", (msg) => {
+        if (msg.data.ampAnalyticsResponse) {
+          processResponse(msg.data.ampAnalyticsResponse);
+        }
+      });
+    }
+  }
+
+  /**
+   * Called when a creative no longer needs its cross-domain iframe (for
+   * instance, because the creative has been removed from the DOM).
+   * Once all creatives using a frame are done with it, the frame can be
+   * destroyed.
+   * @param {!HTMLDocument} ampDoc The AMP document
+   * @param {!Object<string,string>} transportOptions The 'transport' portion
+   * of the amp-analytics config object
+   */
+  static doneWithCrossDomainIframe(ampDoc, transportOptions) {
+    const frameUrl = Transport.appendHashToUrl_(transportOptions['iframe'],
+        transportOptions['dataHash']);
+    if (Transport.hasCrossDomainFrame(frameUrl)) {
+      if (this.decrementUsageCount_(frameUrl) <= 0) {
+        ampDoc.body.removeChild(Transport.crossDomainFrames[frameUrl].frame);
+        delete Transport.crossDomainFrames[frameUrl];
+      }
+    }
+  }
+
+  /**
+   * Returns whether a url of a cross-domain frame is already known
+   * @param {!string} frameUrl
+   * @return {!boolean}
+   */
+  static hasCrossDomainFrame(frameUrl) {
+    return hasOwn(Transport.crossDomainFrames, frameUrl);
+  }
+
+  /**
+   * Create a cross-domain iframe for third-party vendor anaytlics
+   * @param {!HTMLDocument} ampDoc  The document node of the parent page
+   * @param {!string} frameUrl  The URL, including data hash if
+   * applicable, of the cross-domain iframe
+   * @return {!Element}
+   */
+  static createCrossDomainFrame(ampDoc, frameUrl) {
+    // DO NOT MERGE THIS
+    // Warning: the scriptSrc URL below is only temporary. Don't merge
+    // before getting resolution on that.
+    const frame = createElementWithAttributes(ampDoc, 'iframe', {
+      sandbox: 'allow-scripts',
+      name: JSON.stringify({
+        'scriptSrc': '/examples/analytics-3p-remote-frame-helper.js',
+      }),
+    });
+    loadPromise(frame).then(() => {
```

Comment:
I'm trying to remember the flow... were we expecting load to indicate when to send messages or do we expect the library within frame to send a post message?  If the frame contents require numerous resources in addition to the library and/or do not properly include the library, then the post message method is ideal as it would potentially allow communication before all resources are loaded and ensure we do not send messages for an invalid frame.  On the other hand, if the frame contents are small and load often fires before the time it takes to send/receive a post message, the load is better.  Its also possible the frame could load the library AFTER load fires on the frame which is why I'm partial to waiting for a post message indicating its ready.